local-volume/provisioner/pkg/deleter/blockcleaner.go

```diff
@@ -0,0 +1,247 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deleter
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/external-storage/local-volume/provisioner/pkg/common"
+	"k8s.io/api/core/v1"
+	"os/exec"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	// ZeroDev is refers to zero device.
+	ZeroDev = "/dev/zero"
+	// DefaultBlockSize is the default block size to use in dd
+	DefaultBlockSize = "4096"
+	// ShredCommand the shred cli.
+	ShredCommand = "shred"
+	// MkfsCommand is the mkfs executable.
+	MkfsCommand = "mkfs"
+	// DDCommand is the dd executable.
+	DDCommand = "dd"
+	// WipeComand is the dd executable.
+	WipeComand = "wipefs"
+	// ShellCommand is the command to invoke shell
+	ShellCommand = "sh"
+
+	// DDFullErrExpected Error ouptput expected when running dd to fill a device
+	DDFullErrExpected = "No space left on device"
+	// NoErrExpected No error expected from running command
+	NoErrExpected = ""
+)
+
+// BlockCleaner Interface for invoking async script to do cleanup of block device.
+type BlockCleaner interface {
+	// DoCleanup deletes
+	DeleteBlockPV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error
+	isRunning(pvName string) bool
+	markRunning(pvName string)
+	markDone(pvName string)
+}
+
+var _ BlockCleaner = &BlockCleanerImpl{}
+
+// BlockCleanerImpl Implementation of BLockCleaner interface
+type BlockCleanerImpl struct {
+	*common.RuntimeConfig
+	mutex     sync.Mutex
+	procTable map[string]time.Time
+}
+
+// NewBlockCleaner returns a BlockCleaner
+func NewBlockCleaner(config *common.RuntimeConfig) *BlockCleanerImpl {
+	return &BlockCleanerImpl{RuntimeConfig: config,
+		procTable: make(map[string]time.Time),
+	}
+}
+
+func (v *BlockCleanerImpl) isRunning(pvName string) bool {
```

Comment:
What happens if the process restarts and we lose the in-memory state?  Is it safe to run the command twice?

Reply:
Yes. Repeating it would only result in idempotent behavior (except in random, where the output would be a different set of random values, something that is desirable). It would repeat the exact same process and finish the job.